From e1ae07302b9b0b837116ed0d4d2a7326111afc54 Mon Sep 17 00:00:00 2001
From: pedro311 <pedro311@gmail.com>
Date: Tue, 21 Jul 2026 18:08:05 +0000
Subject: [PATCH 1/3] ntpd: schedule retry after initial DNS lookup failure
 (patch)

resolve_peer_hostname() already maintains an exponential dns_errors
counter and the main loop already uses HOSTNAME_INTERVAL for later
lookups. The initial lookup in add_peers(), however, left
next_action_time due immediately when resolution failed.

Schedule the first retry with the existing hostname backoff and do not
consume the initial burst countdown while a peer has no resolved
address.
---
 busybox/networking/ntpd.c | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git busybox/networking/ntpd.c busybox/networking/ntpd.c
index 6ed6491..920da6b 100644
--- busybox/networking/ntpd.c
+++ busybox/networking/ntpd.c
@@ -821,6 +821,11 @@ add_peers(const char *s, key_entry_t *key_entry)
 				return;
 			}
 		}
+	} else {
+		/* Do not leave an unresolved peer immediately due forever.
+		 * The main loop will retry it after this backoff expires.
+		 */
+		set_next(p, HOSTNAME_INTERVAL * p->dns_errors);
 	}
 
 	IF_FEATURE_NTP_AUTH(p->key_entry = key_entry;)
@@ -2596,13 +2601,18 @@ int ntpd_main(int argc UNUSED_PARAM, char **argv)
 
 			if (p->next_action_time <= G.cur_time) {
 				if (p->p_fd == -1) {
-					/* Time to send new req */
-					if (--cnt == 0) {
-						VERB4 bb_simple_error_msg("disabling burst mode");
-						G.polladj_count = 0;
-						G.poll_exp = MINPOLL;
+					/* Do not consume the initial burst countdown until
+					 * hostname resolution has produced an address.
+					 */
+					if (p->p_lsa) {
+						/* Time to send new req */
+						if (cnt != 0 && --cnt == 0) {
+							VERB4 bb_simple_error_msg("disabling burst mode");
+							G.polladj_count = 0;
+							G.poll_exp = MINPOLL;
+						}
+						send_query_to_peer(p);
 					}
-					send_query_to_peer(p);
 				} else {
 					/* Timed out waiting for reply */
 					close(p->p_fd);
-- 
2.47.3

